SIMSBIOHUB-379: Parse intake key-value pairs

curtisupshall opened the pull request:
# Overview

## Links to Jira tickets

 - [SIMSBIOHUB-379](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-379)

## Description of relevant changes

- Adds service and repo methods to parse search fields from submission features

curtisupshall committed:
Squashed SIMSBIOHUB-365

curtisupshall committed:
SIMSBIOHUB-379: initialize search-index repo

curtisupshall committed:
SIMSBIOHUB-379: Finished typedefs for search-index-repo insertion methods

curtisupshall committed:
Merge branch 'dev' into SIMSBIOHUB-379

curtisupshall committed:
Merge branch 'SIMSBIOHUB-379' of github.com:bcgov/biohubbc-platform into SIMSBIOHUB-379

curtisupshall committed:
SIMSBIOHUB-379: Create search indexing endpoint

curtisupshall committed:
SIMSBIOHUB-379: Parse key-value pairs (WIP)

NickPhura commented on api/src/services/search-index-service.ts:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  async indexFeaturesBySubmissionId(submissionId: number): Promise<void> {
+    const submissionService = new SubmissionRepository(this.connection);
```
nit: `submissionRepository`

NickPhura commented on api/src/services/search-index-service.ts:
```diff
@@ -0,0 +1,62 @@
+    this.searchIndexRepository = new SearchIndexRepository(connection);
+  }
+
+  async indexFeaturesBySubmissionId(submissionId: number): Promise<void> {
```
Looks good! 

We will probably have to consider nested properties eventually, where the property feature type is 'array' or 'object. But I think we can just assume the properties are flat for now, in order to get the first pass out the door.

curtisupshall committed:
SIMSBIOHUB-379: fix typo

curtisupshall committed:
Merge branch 'dataset_security_feature' into SIMSBIOHUB-379

curtisupshall committed:
SIMSBIOHUB-379: Add records to table

curtisupshall committed:
Merge branch 'dataset_security_feature' into SIMSBIOHUB-379

curtisupshall committed:
SIMSBIOHUB-379: Restore admin submissions folder

curtisupshall committed:
SIMSBIOHUB-379: some type changes, debugging

curtisupshall committed:
SIMSBIOHUB-379: Added timestamp to db seed

curtisupshall committed:
SIMSBIOHUB-379: Added some tests WIP

curtisupshall committed:
Merge branch 'dataset_security_feature' into SIMSBIOHUB-379

curtisupshall committed:
SIMSBIOHUB-379: Tests WIP

curtisupshall committed:
SIMSBIOHUB-379: Refined types, jsdocs

curtisupshall committed:
SIMSBIOHUB-379: Added repo tests

curtisupshall committed:
SIMSBIOHUB-379: Working tests

curtisupshall committed:
SIMSBIOHUB-379: Code cleanup

curtisupshall committed:
Merge branch 'dataset_security_feature' into SIMSBIOHUB-379

curtisupshall committed:
Merge branch 'dataset_security_feature' into SIMSBIOHUB-379

curtisupshall committed:
SIMSBIOHUB-379: Fix type error

al-rosenthal committed:
Merge branch 'dataset_security_feature' into SIMSBIOHUB-379

NickPhura committed:
Update migration

NickPhura committed:
Update intake endpoint to work with related changes from SIMS PR #1185
Update validation to account for intake endpoint changes (enhance to support validating any level of nested submission features)
Fix bug with zod geo json schemas.
Update unit tests.

NickPhura commented:
@curtisupshall I pushed some changes into this PR, mostly just around updating the intake endpoint to work with the latest structure being sent from SIMS (SIMS changes are in this open PR https://github.com/bcgov/biohubbc/pull/1185). Once the SIMS PR goes through, we should be able to run the sims publish and see the BioHub intake get hit, including the validation and a call to your parsing code.

NickPhura committed:
Fix unit test

NickPhura committed:
Fix intake validation

NickPhura commented on api/src/services/search-index-service.ts:
```diff
@@ -0,0 +1,97 @@
+    });
+
+    if (datetimeRecords.length) {
+      this.searchIndexRepository.insertSearchableDatetimeRecords(datetimeRecords);
```
These need `awaits`

curtisupshall replied:
Good catch

NickPhura commented on api/src/services/search-index-service.ts:
```diff
@@ -0,0 +1,97 @@
+
+    features.forEach((feature) => {
+      const { submission_feature_id } = feature;
+      Object.entries(feature.data.properties).forEach(([feature_property_name, value]) => {
```
Should just be `feature.data` rather than `feature.data.properties`

curtisupshall replied:
Fixed

NickPhura commented on api/src/repositories/search-index-respository.ts:
```diff
@@ -0,0 +1,222 @@
+
+    const queryBuilder = getKnex().queryBuilder().insert(spatialRecords).into('search_spatial').returning('*');
+
+    const response = await this.connection.knex<SpatialSearchableRecord>(queryBuilder);
```
The generic type should be the second parameter in the function call: `this.connection.knex(queryBuilder, SpatialSearchableRecord);`

The way it is now, it just forces the result type without actually running the zod check.

curtisupshall replied:
Good catch, fixed.

NickPhura commented on api/src/repositories/search-index-respository.ts:
```diff
@@ -0,0 +1,222 @@
+
+    const queryBuilder = getKnex().queryBuilder().insert(spatialRecords).into('search_spatial').returning('*');
+
+    const response = await this.connection.knex<SpatialSearchableRecord>(queryBuilder);
```
Got an error when the code reached here: the issue is that the incoming geojson data is being inserted into the `geometry` column, which is the postgis geometry type column. 

The raw geojson would need to go in the `geojson` column, and then we also need to populate the `geometry` column with the encoded value. Can see similar examples of converting raw geojson into postgis geometry if you search for `ST_GeomFromGeoJSON`.

NickPhura replied:
Just realized/starting to remember that we actually can't convert GeoJSON Feature or FeatureCollection types to postgis geometry. It only accepts regular type (polygon, point, etc) and GeometryCollection. We actually have a function in SIMS `generateGeometryCollectionSQL` for this exact reason. So we either need to steal it or create a better version and leverage that here.

curtisupshall replied:
Solved.

curtisupshall committed:
SIMSBIOHUB-379: Addressed PR feedback

curtisupshall committed:
SIMSBIOHUB-379: Fix zod schema checking for index table insertion

curtisupshall committed:
SIMSBIOHUB-379: Promise typing

curtisupshall committed:
Merge branch 'dataset_security_feature' into SIMSBIOHUB-379

curtisupshall committed:
SIMSBIOHUB-379: Fixed geometry typing

NickPhura committed:
Improve JSONPath string in submission validation-service.ts

NickPhura committed:
Merge branch 'SIMSBIOHUB-379' of https://github.com/bcgov/biohubbc-platform into SIMSBIOHUB-379

al-rosenthal reviewed (commented):
Running this locally I've run into an issue publishing a survey without an end date. The UI doesn't force a survey end date to be specified but the intake in BioHub fails when trying to insert into the `search_datetime` table. Should end date be required in SIMS? will that be applicable for other submitters?

curtisupshall commented:
> Running this locally I've run into an issue publishing a survey without an end date. The UI doesn't force a survey end date to be specified but the intake in BioHub fails when trying to insert into the `search_datetime` table. Should end date be required in SIMS? will that be applicable for other submitters?

Good catch! Committed a change which ignores date values that are `null`.

@NickPhura thoughts on this approach?

curtisupshall committed:
SISMBIOHUB-379: Filtered out null end_date

KjartanE commented on api/src/paths/submission/search-idx.ts:
```diff
@@ -0,0 +1,85 @@
+      description: 'TODO', // TODO
+      content: {
+        'application/json': {
+          schema: {
```
api doc still required todo

KjartanE commented on app/src/hooks/api/useDatasetApi.ts:
```diff
@@ -32,6 +32,19 @@ const useDatasetApi = (axios: AxiosInstance) => {
+   */
+  const getDataset = async (datasetUUID: string): Promise<any> => {
+    const { data } = await axios.get(`api/dataset/${datasetUUID}`);
+    console.log('data', data);
```
nit: remove

curtisupshall replied:
Removed

KjartanE commented on api/src/services/artifact-service.ts:
```diff
@@ -84,7 +84,11 @@ export class ArtifactService extends DBService {
-    const { submission_id } = await this.submissionService.insertSubmissionRecordWithPotentialConflict(dataPackageId);
+    const { submission_id } = await this.submissionService.insertSubmissionRecordWithPotentialConflict(
+      dataPackageId,
+      'TODO_Temp',
```
Is this function being used? could we delete the entire function?

NickPhura replied:
There is a lot of old code we can remove. We were just holding off so as no to make actual feature PRs bigger/more complex.

NickPhura replied:
That being said, removing small bits as needed, if it is hindering progress on current work, is totally fine.

KjartanE commented on api/src/paths/submission/intake.ts:
```diff
@@ -40,26 +42,28 @@ POST.apiDoc = {
               items: {
                 $ref: '#/components/schemas/SubmissionFeature'
               },
+              maxItems: 1,
```
why maxItems? This endpoint should accept all features connected to submission

curtisupshall committed:
SIMSBIOHUB-379: Remove console log

al-rosenthal reviewed (approved):
running this locally and everything looks good 🕶️

curtisupshall committed:
Merge branch 'dataset_security_feature' into SIMSBIOHUB-379

curtisupshall committed:
SIMSBIOHUB-379: Lint fixes

curtisupshall committed:
SIMSBIOHUB-379: Added test check

curtisupshall committed:
SIMSBIOHUB-379: Fix compilation errors in tests

curtisupshall committed:
SIMSBIOHUB-379: Fixed test

NickPhura committed:
Remove .only

curtisupshall committed:
SIMSBIOHUB-379: Fixed typo in test

curtisupshall committed:
SIMSBIOHUB-379: Added missing stub for intake endpoint test

curtisupshall committed:
Merge branch 'dataset_security_feature' into SIMSBIOHUB-379

curtisupshall committed:
SIMSBIOHUB-379: Ran linter
